ksqldb-engine/src/main/java/io/confluent/ksql/execution/pull/PullPhysicalPlan.java

```diff
@@ -79,27 +84,53 @@ public PullPhysicalPlan(
 
   public void execute(
       final List<KsqlPartitionLocation> locations,
-      final PullQueryQueue pullQueryQueue,
-      final BiFunction<List<?>, LogicalSchema, PullQueryRow> rowFactory) {
-
+      final PullQueryWriteStream pullQueryQueue,
+      final Function<StreamedRow, StreamedRow> addDebugInfo
+  ) {
     // We only know at runtime which partitions to get from which node.
     // That's why we need to set this explicitly for the dataSource operators
     dataSourceOperator.setPartitionLocations(locations);
 
+    // to achieve parity with the remote execution path, we start
+    // by passing the row header into the queue
+    pullQueryQueue.write(ImmutableList.of(StreamedRow.header(queryId, schema)));
+
     open();
-    QueryRow row;
-    while ((row = (QueryRow)next()) != null) {
-      if (pullQueryQueue.isClosed()) {
-        // If the queue has been closed, we stop adding rows and cleanup. This should be triggered
-        // because the client has closed their connection with the server before the results have
-        // completed.
-        LOGGER.info("Queue closed before results completed. Stopping execution.");
-        break;
-      }
-      if (!pullQueryQueue.acceptRow(rowFactory.apply(row.value().values(), schema))) {
-        LOGGER.info("Failed to queue row");
+    QueryRow row = (QueryRow) next();
+
+    int rowsQueued = 0;
+    while (!pullQueryQueue.isDone() && row != null) {
+      // this mechanism is "best effort" in that it is possible that multiple
+      // threads waiting on this condition may all be signalled and attempt
+      // to write into that queue - this is OK as it is just a mechanism for
+      // backpressure (a soft limit)
+      if (pullQueryQueue.writeQueueFull()) {
+        try {
+          if (!pullQueryQueue.awaitCapacity(100, TimeUnit.MILLISECONDS)) {
```

Comment:
So....  something somewhere could get an extra 100 ms worth of waiting, right?

What situations would see that?

Reply:
yes this is correct, though it was also the behavior of `PullQueryQueue` https://github.com/confluentinc/ksql/blob/master/ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryQueue.java#L255-L260

Basically this is the case when a connection is terminated or the limit is otherwise hit while the queue is full. It's pretty unlikely as we almost always flush the queue.